Use opaque id in task cancellation assertion

mhl-b opened the pull request:
Add use of Opaque ID HTTP header in task cancellation assertion. In some tests, like this #88201 `testCatSegmentsRestCancellation`, we assert that all tasks related to specific HTTP request are cancelled. But we do blanket approach in assertion block catching all tasks by action name. I think narrowing down assertion to specific http request in this case would be more accurate.

It is still not clear why test mentioned above failing, but after hours of investigation and injecting random delays, I'm inclining more to @DaveCTurner's comment about interference from other tests or cluster activity. I added additional log that will report when we spot task with different opaque id.

mhl-b committed:
narrow-down tasks assertion per http request

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

mhl-b committed:
cleanup

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/test/TaskAssertions.java:
```diff
@@ -70,6 +72,14 @@ public static void assertAllCancellableTasksAreCancelled(String actionPrefix) th
                 for (CancellableTask cancellableTask : taskManager.getCancellableTasks().values()) {
                     if (cancellableTask.getAction().startsWith(actionPrefix)) {
                         logger.trace("--> found task with prefix [{}]: [{}]", actionPrefix, cancellableTask);
+                        if (taskOpaqueIdMatches(cancellableTask.headers(), opaqueId) == false) {
```
I would rather not leniently skip non-matching tasks: we shouldn't have any such tasks in the first place. If this is the problem then I'd prefer to track down where those tasks are coming from and get rid of them at source.

Let's assert that the task ID matches here instead.

DaveCTurner commented on qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/BlockedSearcherRestCancellationTestCase.java:
```diff
@@ -71,11 +73,32 @@ protected boolean addMockInternalEngine() {
+    // returns new or existing id
+    private static String setOpaqueIdIfNotExists(Request request) {
+        String opaqueId;
+        if (request.getOptions().containsHeader(Task.X_OPAQUE_ID_HTTP_HEADER)) {
```
I think this is dead code today since none of the requests include an opaque-id header. Let's just assert it's absent.

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/test/TaskAssertions.java:
```diff
@@ -83,11 +93,31 @@ public static void assertAllCancellableTasksAreCancelled(String actionPrefix) th
+    }
+
+    private static boolean taskOpaqueIdMatches(Map<String, String> taskHeaders, @Nullable String opaqueId) {
+        if (opaqueId == null) {
```
I think we should assert the opaque ID is missing if we didn't expect to see one.

DaveCTurner reviewed (commented):
Seems like a good idea to me. I left some inline comments mainly about checking expectations more strictly.

mhl-b committed:
add assertions

mhl-b committed:
more tracing and error logs

mhl-b committed:
spotless

mhl-b commented:
@DaveCTurner thank you for feedback. I added assertions and test names to the opaque id. Also added more explicit exception message in `assertAllCancellableTasksAreCancelled` that will show all tasks rather than the one that does not pass.

DaveCTurner commented on qa/smoke-test-http/src/javaRestTest/java/org/elasticsearch/http/BlockedSearcherRestCancellationTestCase.java:
```diff
@@ -71,11 +72,16 @@ protected boolean addMockInternalEngine() {
     }
 
-    void runTest(Request request, String actionPrefix) throws Exception {
+    void runTest(String testName, Request request, String actionPrefix) throws Exception {
```
Rather than passing in a string that (hopefully) identifies the test, I'd suggest using `org.apache.lucene.tests.util.LuceneTestCase#getTestName` (and maybe `org.apache.lucene.tests.util.LuceneTestCase#getTestClass`).

DaveCTurner commented on server/src/main/java/org/elasticsearch/tasks/CancellableTask.java:
```diff
@@ -19,9 +19,9 @@
  */
 public class CancellableTask extends Task {
 
+    private final SubscribableListener<Void> listeners = new SubscribableListener<>();
```
Nit: did you mean to move this field?

mhl-b replied:
applied formatting to entire file...

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/test/TaskAssertions.java:
```diff
@@ -59,35 +62,34 @@ private static void awaitTaskWithPrefix(String actionPrefix, Iterable<TransportS
             assertTrue(tasks.toString(), tasks.stream().noneMatch(t -> t.action().startsWith(actionPrefix)));
         });
     }
+
```
nit: extra line?

DaveCTurner reviewed (commented):
Looks good, one suggestion about how to identify the test more neatly

mhl-b committed:
feedback

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into assert-task-per-request

DaveCTurner reviewed (approved):
LGTM

mhl-b committed:
poke ci
